site/docs/spec/terminology.yaml

```diff
@@ -0,0 +1,230 @@
+  define {function}s, {type class}es, and {type variation}s.
+advanced extension: >-
+  an {extension} to Substrait, used within a plan by populating one of the
+  various protobuf `Any` messages contained therein.
+
+anchor: >-
+  an integer in the protobuf serialization that can be referred to by
+  means of {reference}s elsewhere in a {plan}. Anchors are used for
+  {simple extension} URIs, {user-defined} {function}s, {user-defined}
+  {type class}es, and {user-defined} {type variation}s. Anchors are namespaced
+  by their type, but must otherwise be unique within the plan. Their value has
+  no significance other than to be a unique thing to refer to.
+reference: >-
+  an integer in the protobuf serialization that can be used to refer to
+  an {anchor}.
+
+function: >-
+  some {user-defined} operation on some information (usually specified by means
+  of {function argument}s) that yields a single {datum} for each invocation.
+  Functions are declared via {function declaration}s in {simple extension}s,
+  and referred to from within a plan via {function binding}s. They are
+  subdivided into {scalar function}s, {aggregate function}s, and
+  {window function}s.
+scalar function: >-
+  a {function} that computes a {datum} for each row of a {dataset} individually
+  within {expression} context, as defined
+  [here](../expressions/scalar_functions.md).
+aggregate function: >-
+  a {function} that computes a single {datum} for a complete {dataset}, as
+  defined [here](../expressions/aggregate_functions.md).
+window function: >-
+  a {function} that computes a {datum} for each row of a {dataset}, making use
+  of some window of surrounding rows in addition to the row it is computed for,
+  as defined [here](../expressions/window_functions.md).
+
+function declaration: >-
+  information about a {function}, defined in a {simple extension}.
+function parameter: >- # changed from "argument" and in some cases "variable"
+  a {static}, named slot which a {function argument} can be bound. Part of a
+  {function declaration}. Can be a {value function parameter},
+  {type function parameter}, or {enum function parameter}.
+value function parameter: >- # changed from "value argument"
+  a {function parameter} to which a {value function argument} can be bound.
+  The {data type} of the argument must match the {data type pattern} defined
+  along with the parameter.
+type function parameter: >- # changed from "type argument"
+  a {function parameter} to which a {type function argument} can be bound.
+  The {data type} passed to the argument must match the {data type pattern}
+  defined along with the parameter.
+enum function parameter: >- # changed from "enumeration"
+  a {function parameter} to which an {enum function argument} can be bound.
+  The allowed values are defined along with the parameter by means of a set
+  of string options and an optional default option.
+
+function binding: >- # didn't really exist
```

Comment:
Those are fundamentally different things. Binding happens during compilation, and involves connecting the parameters in the function definition (or actually declaration/prototype in this case, since Substrait doesn't know the function definition) to argument expressions to build a complete expression tree. Invocation (aka calling a function) is the process of actually executing the body of a function to evaluate its return value, once all its dependencies (argument expressions and/or program state for impure functions) have been evaluated, either at runtime or during constant propagation. After binding, an expression tree with a function binding in it may be modified during optimization, but the bound function generally can't be invoked yet. The relationships are `definition 1-->* binding 1-->* invocation`.

FWIW, in the context you linked the difference between "binding" and "invocation" is like "addition" and "add" (if anything), so I don't think the distinction is important enough to require a refactor there, in the same way that you can get away with mixing "parameter" and "argument" in 99% of cases. It's the 1% that I'm trying to cover.